cupy/cuda/cupy_thrust.cu

```diff
@@ -109,37 +109,85 @@ template void cupy::thrust::_lexsort<cpy_double>(size_t *, void *, size_t, size_
  */
 
 template <typename T>
-void cupy::thrust::_argsort(size_t *idx_start, void *data_start, size_t num) {
+void cupy::thrust::_argsort(size_t *idx_start, void *data_start, void *keys_start, void *buff_start, const std::vector<ptrdiff_t>& shape) {
     /* idx_start is the beggining of the output array where the indexes that
        would sort the data will be placed. The original contents of idx_start
        will be destroyed. */
 
-    device_ptr<T> dp_data_first, dp_data_last;
+    size_t ndim = shape.size();
+    ptrdiff_t size;
+
     device_ptr<size_t> dp_idx_first, dp_idx_last;
+    device_ptr<T> dp_data_first, dp_data_last;
+    device_ptr<size_t> dp_keys_first, dp_keys_last;
+    device_ptr<T> dp_buff_first, dp_buff_last;
+
+    // Compute the total size of the data array.
+    size = shape[0];
+    for (size_t i = 1; i < ndim; ++i) {
+        size *= shape[i];
+    }
 
     // Cast device pointers of data.
     dp_data_first = device_pointer_cast(static_cast<T*>(data_start));
-    dp_data_last  = device_pointer_cast(static_cast<T*>(data_start) + num);
+    dp_data_last  = device_pointer_cast(static_cast<T*>(data_start) + size);
 
     // Generate an index sequence.
     dp_idx_first = device_pointer_cast(static_cast<size_t*>(idx_start));
-    dp_idx_last  = device_pointer_cast(static_cast<size_t*>(idx_start) + num);
-    sequence(dp_idx_first, dp_idx_last);
-
-    // Sort the index sequence by data.
-    stable_sort_by_key(dp_data_first,
-                       dp_data_last,
-                       dp_idx_first,
-                       less<T>());
+    dp_idx_last  = device_pointer_cast(static_cast<size_t*>(idx_start) + size);
+    transform(make_counting_iterator<size_t>(0),
+              make_counting_iterator<size_t>(size),
+              make_constant_iterator<ptrdiff_t>(shape[ndim-1]),
+              dp_idx_first,
+              modulus<size_t>());
+
+    if (ndim == 1) {
+        // Sort the index sequence by data.
+        stable_sort_by_key(dp_data_first,
+                           dp_data_last,
+                           dp_idx_first,
+                           less<T>());
+    } else {
+        // Generate key indices.
+        dp_keys_first = device_pointer_cast(static_cast<size_t*>(keys_start));
+        dp_keys_last  = device_pointer_cast(static_cast<size_t*>(keys_start) + size);
+        transform(make_counting_iterator<size_t>(0),
+                  make_counting_iterator<size_t>(size),
+                  make_constant_iterator<ptrdiff_t>(shape[ndim-1]),
+                  dp_keys_first,
+                  divides<size_t>());
+
+        // Copy data to the buffer.
+        dp_buff_first = device_pointer_cast(static_cast<T*>(buff_start));
+        dp_buff_last  = device_pointer_cast(static_cast<T*>(buff_start) + size);
+        copy(dp_data_first, dp_data_last, dp_buff_first);
+
+        // Sorting with back-to-back approach.
+
+        stable_sort_by_key(dp_buff_first,
```

Comment:
Sorry, this comment is my misunderstanding.
But, Can we use `thrust::make_zip_iterator`?